{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "from pysam import VariantFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vcf = \"/media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/merged_allsites.vcf\"\n",
    "out_vcf = \"/media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/merged_allsites.GETS.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tD2A1\tD2B3\tD2C1\tD2C3\tD3A1\tD3A3\tD4A3\tD4B4\tD5B3\tD5C1\tD5C3\tH2A3\tH2B4\tH2C3\tH3A4\tH3C4\tH4A4\tH4C2\tH5A2\tH5A3\tH5A4\tH5C2\tancestor\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head -39 /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/merged_allsites.vcf | tail -1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Micro-evolution experiment\n",
    "\n",
    "## Create new VCF with SH, TP, GET infos\n",
    "\n",
    "### 100 generations endpoint only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##ALT=<ID=NON_REF,Description=\"Represents any possible alternative allele not already represented at this location by REF and ALT\">\n",
      "##FILTER=<ID=LowQual,Description=\"Low quality\">\n",
      "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=PGT,Number=1,Type=String,Description=\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles\">\n",
      "##FORMAT=<ID=PID,Number=1,Type=String,Description=\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\">\n",
      "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "##FORMAT=<ID=PS,Number=1,Type=Integer,Description=\"Phasing set (typically the position of the first variant in the set)\">\n",
      "##FORMAT=<ID=RGQ,Number=1,Type=Integer,Description=\"Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong)\">\n",
      "##FORMAT=<ID=SB,Number=4,Type=Integer,Description=\"Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.\">\n",
      "##GATKCommandLine=<ID=GenomicsDBImport,CommandLine=\"GenomicsDBImport --genomicsdb-workspace-path /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/combined.db --batch-size 10 --sample-name-map /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/sample.map.txt --max-num-intervals-to-import-in-parallel 4 --intervals /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/endpoint/intervals.list --reference /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/input/reference.fa --genomicsdb-segment-size 1048576 --genomicsdb-vcf-buffer-size 16384 --overwrite-existing-genomicsdb-workspace false --consolidate false --validate-sample-name-map false --merge-input-intervals false --reader-threads 1 --merge-contigs-into-num-partitions 0 --genomicsdb-shared-posixfs-optimizations false --genomicsdb-use-gcs-hdfs-connector false --interval-set-rule UNION --interval-padding 0 --interval-exclusion-padding 0 --interval-merging-rule ALL --read-validation-stringency SILENT --seconds-between-progress-updates 10.0 --disable-sequence-dictionary-validation false --create-output-bam-index true --create-output-bam-md5 false --create-output-variant-index true --create-output-variant-md5 false --max-variants-per-shard 0 --lenient false --add-output-sam-program-record true --add-output-vcf-command-line true --cloud-prefetch-buffer 0 --cloud-index-prefetch-buffer 0 --disable-bam-index-caching false --sites-only-vcf-output false --help false --version false --showHidden false --verbosity INFO --QUIET false --use-jdk-deflater false --use-jdk-inflater false --gcs-max-retries 20 --gcs-project-for-requester-pays  --disable-tool-default-read-filters false\",Version=\"4.2.1.0\",Date=\"August 28, 2023 8:13:38 AM CEST\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">\n",
      "##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">\n",
      "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
      "##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">\n",
      "##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">\n",
      "##INFO=<ID=MQRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\">\n",
      "##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant Confidence/Quality by Depth\">\n",
      "##INFO=<ID=RAW_MQandDP,Number=2,Type=Integer,Description=\"Raw data (sum of squared MQ and total depth) for improved RMS Mapping Quality calculation. Incompatible with deprecated RAW_MQ formulation.\">\n",
      "##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\">\n",
      "##INFO=<ID=SOR,Number=1,Type=Float,Description=\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\">\n",
      "##contig=<ID=Chrom_3,length=20354777>\n",
      "##contig=<ID=Chrom_1,length=18146847>\n",
      "##contig=<ID=Chrom_5,length=16930519>\n",
      "##contig=<ID=Chrom_2,length=16274841>\n",
      "##contig=<ID=Chrom_4,length=15224634>\n",
      "##contig=<ID=Chrom_6,length=13893210>\n",
      "##source=GenomicsDBImport\n",
      "##FORMAT=<ID=GET,Number=1,Type=String,Description=\"Type of variation compared to the ancestor sample.\">\n",
      "##FORMAT=<ID=SH,Number=1,Type=Integer,Description=\"Number of samples sharing the same variant as the considered sample.\">\n",
      "##FORMAT=<ID=TP,Number=1,Type=String,Description=\"Type of variant.\">\n",
      "##INFO=<ID=SH,Number=R,Type=Integer,Description=\"Number of samples sharing the same allele.\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tD2A1\tD2B3\tD2C1\tD2C3\tD3A1\tD3A3\tD4A3\tD4B4\tD5B3\tD5C1\tD5C3\tH2A3\tH2B4\tH2C3\tH3A4\tH3C4\tH4A4\tH4C2\tH5A2\tH5A3\tH5A4\tH5C2\tancestor\n",
      "\n",
      "Elapsed records: 0\n",
      "Elapsed records: 1000000\n",
      "Elapsed records: 2000000\n",
      "Elapsed records: 3000000\n",
      "Elapsed records: 4000000\n",
      "Elapsed records: 5000000\n",
      "Elapsed records: 6000000\n",
      "Elapsed records: 7000000\n",
      "Elapsed records: 8000000\n",
      "Elapsed records: 9000000\n",
      "Elapsed records: 10000000\n",
      "Elapsed records: 11000000\n",
      "Elapsed records: 12000000\n",
      "Elapsed records: 13000000\n",
      "Elapsed records: 14000000\n",
      "Elapsed records: 15000000\n",
      "Elapsed records: 16000000\n",
      "Elapsed records: 17000000\n",
      "Elapsed records: 18000000\n",
      "Elapsed records: 19000000\n",
      "Elapsed records: 20000000\n",
      "Elapsed records: 21000000\n",
      "Elapsed records: 22000000\n",
      "Elapsed records: 23000000\n",
      "Elapsed records: 24000000\n",
      "Elapsed records: 25000000\n",
      "Elapsed records: 26000000\n",
      "Elapsed records: 27000000\n",
      "Elapsed records: 28000000\n",
      "Elapsed records: 29000000\n",
      "Elapsed records: 30000000\n",
      "Elapsed records: 31000000\n",
      "Elapsed records: 32000000\n",
      "Elapsed records: 33000000\n",
      "Elapsed records: 34000000\n",
      "Elapsed records: 35000000\n",
      "Elapsed records: 36000000\n",
      "Elapsed records: 37000000\n",
      "Elapsed records: 38000000\n",
      "Elapsed records: 39000000\n",
      "Elapsed records: 40000000\n",
      "Elapsed records: 41000000\n",
      "Elapsed records: 42000000\n",
      "Elapsed records: 43000000\n",
      "Elapsed records: 44000000\n",
      "Elapsed records: 45000000\n",
      "Elapsed records: 46000000\n",
      "Elapsed records: 47000000\n",
      "Elapsed records: 48000000\n",
      "Elapsed records: 49000000\n",
      "Elapsed records: 50000000\n",
      "Elapsed records: 51000000\n",
      "Elapsed records: 52000000\n",
      "Elapsed records: 53000000\n",
      "Elapsed records: 54000000\n",
      "Elapsed records: 55000000\n",
      "Elapsed records: 56000000\n",
      "Elapsed records: 57000000\n",
      "Elapsed records: 58000000\n",
      "Elapsed records: 59000000\n",
      "Elapsed records: 60000000\n",
      "Elapsed records: 61000000\n",
      "Elapsed records: 62000000\n",
      "Elapsed records: 63000000\n",
      "Elapsed records: 64000000\n",
      "Elapsed records: 65000000\n",
      "Elapsed records: 66000000\n",
      "Elapsed records: 67000000\n",
      "Elapsed records: 68000000\n",
      "Elapsed records: 69000000\n",
      "Elapsed records: 70000000\n",
      "Elapsed records: 71000000\n",
      "Elapsed records: 72000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed records: 73000000\n",
      "Elapsed records: 74000000\n",
      "Elapsed records: 75000000\n",
      "Elapsed records: 76000000\n",
      "Elapsed records: 77000000\n",
      "Elapsed records: 78000000\n",
      "Elapsed records: 79000000\n",
      "Elapsed records: 80000000\n",
      "Elapsed records: 81000000\n",
      "Elapsed records: 82000000\n",
      "Elapsed records: 83000000\n",
      "Elapsed records: 84000000\n",
      "Elapsed records: 85000000\n",
      "Elapsed records: 86000000\n",
      "Elapsed records: 87000000\n",
      "Elapsed records: 88000000\n",
      "Elapsed records: 89000000\n",
      "Elapsed records: 90000000\n",
      "Elapsed records: 91000000\n",
      "Elapsed records: 92000000\n",
      "Elapsed records: 93000000\n",
      "Elapsed records: 94000000\n",
      "Elapsed records: 95000000\n",
      "Elapsed records: 96000000\n",
      "Elapsed records: 97000000\n",
      "Elapsed records: 98000000\n",
      "Elapsed records: 99000000\n",
      "Elapsed records: 100000000\n",
      "CPU times: user 10h 5min 26s, sys: 3min 23s, total: 10h 8min 50s\n",
      "Wall time: 10h 10min 5s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "vcf_in = VariantFile(in_vcf)  # auto-detect input format\n",
    "vcf_out = VariantFile(out_vcf, 'w', header=vcf_in.header)\n",
    "\n",
    "# For writing out in a new VCF file\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_out.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele.\")])\n",
    "\n",
    "# Necessary to allow setting new tags in the variant records and record samples\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_in.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele (ref or alt).\")])\n",
    "\n",
    "# DEBUG : check if header was modified indeed\n",
    "print(vcf_out.header)\n",
    "\n",
    "ancestor = \"ancestor\"\n",
    "\n",
    "# For record in sample\n",
    "for n, rec in enumerate(vcf_in) :\n",
    "    if n % 1000000 == 0 :\n",
    "        print(\"Elapsed records: {}\".format(n))\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 11196 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    cAlleles = {k:0 for k in rec.alleles} # For INFO SH\n",
    "    allAlleles = {} # Will store tuples for each sample for later counting\n",
    "    \n",
    "    # For sample in record\n",
    "    for k, sm in rec.samples.items() :\n",
    "        cCurrentAlleles = []\n",
    "            \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 11196 :\n",
    "            print(k, cCurrentAlleles)\n",
    "        \"\"\"\n",
    "           \n",
    "        try : # In case no AD -> go to except\n",
    "            for al, dp in zip(rec.alleles, sm[\"AD\"]) :\n",
    "                # Check each allele if 0.1 < AF < 0.9 or if called by GATK --> add to the cAlleles and cCurrentAlleles\n",
    "                freq = 0.0\n",
    "                try :\n",
    "                    freq = dp/sm[\"DP\"]\n",
    "                except :\n",
    "                    freq = dp/sum(sm[\"AD\"])\n",
    "\n",
    "                if freq >= 0.1 or al in sm.alleles :\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "        except :\n",
    "            try : # In case no DP --> except\n",
    "                if len(set(sm.alleles)) == 1 : # If is homozygous\n",
    "                    al = set(sm.alleles).pop()\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] = 1\n",
    "                    else :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "            except :\n",
    "                print(rec)\n",
    "                print(\"Unexpected error:\", sys.exc_info()[0])\n",
    "                #break\n",
    "                pass\n",
    "\n",
    "        allAlleles[k] = tuple(cCurrentAlleles) # Create a tuple (unchangeable and unsortable --> less mistakes)\n",
    "        # Tuple contains all alleles found in the sample (with 0.1 < AF < 0.9 OR called by GATK)\n",
    "        \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 54641 :\n",
    "            print(k, allAlleles[k])\n",
    "        \"\"\"\n",
    "        \n",
    "        if len(allAlleles[k]) == 0 or None in sm.alleles : # If no alleles found or uncalled genotype\n",
    "            sm[\"TP\"] = \"UC\" # TYPE = Uncalled\n",
    "        else : # At least one allele found\n",
    "            if len(allAlleles[k]) == 1 : # Mono-allelic\n",
    "                if allAlleles[k][0] == rec.ref :\n",
    "                    sm[\"TP\"] = \"RF\" # TYPE = Hom Ref \n",
    "                else :\n",
    "                    if len(allAlleles[k][0]) == 1 and allAlleles[k][0] != \"*\" :\n",
    "                        sm[\"TP\"] = \"AS\" # TYPE = Alternative SNV\n",
    "                    else :\n",
    "                        if allAlleles[k][0] == \"<NON_REF>\" :\n",
    "                            sm[\"TP\"] = \"AU\" # TYPE = Alternative UNKNOWN\n",
    "                        else :\n",
    "                            sm[\"TP\"] = \"AI\" # Alternative INDEL\n",
    "            elif len(allAlleles[k]) == 2 : # TYPE = Bi-allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BS\" # Bi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BU\" # TYPE = Bi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"BI\" # TYPE = Bi-allelic INDEL\n",
    "            else : # Multi-Allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MS\" # TYPE = Multi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MU\" # TYPE = Multi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"MI\" # TYPE = Multi-allelic INDEL\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    try :\n",
    "        Anc = rec.samples[ancestor][\"TP\"]\n",
    "    except :\n",
    "        print(\"NO TP\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "\n",
    "    try :\n",
    "        AncAl = allAlleles[ancestor]\n",
    "    except :\n",
    "        print(\"NO KEYS\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # Add tag GET to each sample\n",
    "    typeDict = {\"UC\":0, \"AU\":0, \"BU\":0, \"MU\":0, \"RF\":1, \"AS\":1, \"AI\": 1, \"BI\":2, \"BS\":2, \"MI\":3, \"MS\":3}\n",
    "    AncAl = allAlleles[ancestor]\n",
    "    Anc = rec.samples[ancestor][\"TP\"]\n",
    "    A = typeDict[Anc]\n",
    "    for k, sm in rec.samples.items() :\n",
    "        if sm.name == ancestor :\n",
    "            continue\n",
    "        else :\n",
    "            OffAl = allAlleles[k]\n",
    "            Off = sm[\"TP\"]\n",
    "            O = typeDict[Off]\n",
    "            \n",
    "            intersect = set(AncAl) & set(OffAl)\n",
    "            NumShared = len(intersect)\n",
    "            AllShared = bool(NumShared == len(set(AncAl))) # If all ancestors alleles are found in the offspring (and vice-versa)\n",
    "            NoneShared = bool(NumShared == 0) # If no common alleles\n",
    "            \n",
    "            if A == 0 or O == 0 :\n",
    "                sm[\"GET\"] = \"U\"\n",
    "            elif A == 1 and O == 1 :\n",
    "                if AllShared : # Same allele --> A or 0\n",
    "                    if Off == \"RF\" : # Ref --> 0\n",
    "                        sm[\"GET\"] = \"0\"\n",
    "                    else : # Not ref --> A\n",
    "                        sm[\"GET\"] = \"A\" \n",
    "                else : # Different allele --> S1\n",
    "                    sm[\"GET\"] = \"S1\"\n",
    "            elif A == 1 and O == 2 : # Offspring heterozygous\n",
    "                if NoneShared : # No common allele --> S4\n",
    "                    sm[\"GET\"] = \"S4\"\n",
    "                else : # 1 shared (cannot be more than one because 2 = biallelic) --> N\n",
    "                    sm[\"GET\"] = \"N\"\n",
    "            elif A == 1 and O == 3 : # Offspring multi-allelic --> always suspect S4\n",
    "                sm[\"GET\"] = \"S4\"\n",
    "            elif A == 2 and O == 1 : # Ancestor heterozygous and offspring homozygous\n",
    "                if NoneShared : # No allele shared --> suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1 allele in common --> cannot be more\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 2 and O == 2 : # both heterozygous\n",
    "                if NoneShared : # No common allele --> S2\n",
    "                    sm[\"GET\"] = \"S2\"\n",
    "                else : \n",
    "                    if NumShared == 1 : # One in common\n",
    "                        sm[\"GET\"] = \"M\"\n",
    "                    else : # Two in common --> identical site <-> AllShared\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "            elif A == 2 and O == 3 : # Ancestor biallelic het and offspring multiallelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S2\" \n",
    "                else :\n",
    "                    if NumShared == 1 : # 1 common allele --> more than 2 events involved : suspect\n",
    "                        sm[\"GET\"] = \"S2\"\n",
    "                    else : # 2 common alleles --> Increased heterozygosity\n",
    "                        sm[\"GET\"] = \"I\"\n",
    "            elif A == 3 and O == 1 : # Ancestor multi-allelic and offspring homozygous\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\" \n",
    "                else : # 1 shared --> lost heterozygosity\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 3 and O == 2 : # Ancestor multi-allelic and offspring bi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    sm[\"GET\"] = \"D\"\n",
    "            elif A == 3 and O == 3 : # Both multi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    if AllShared :\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "                    else :\n",
    "                        if len(AncAl) > len(OffAl) : # More alleles in ancestor : Decreased Het\n",
    "                            sm[\"GET\"] = \"D\"\n",
    "                        else : # More alleles in offspring : Increased Het\n",
    "                            sm[\"GET\"] = \"I\"\n",
    "            else : # Should not come here\n",
    "                # DEBUG\n",
    "                print(n, allAlleles)\n",
    "                print(k, rec)\n",
    "                print(\"----\")\n",
    "                \n",
    "                pass\n",
    "    \n",
    "    # Add UNCALLED samples to the dictionary for later sharedness\n",
    "    for sample in vcf_out.header.samples :\n",
    "        if sample not in allAlleles.keys() :\n",
    "            allAlleles[sample] = tuple([])\n",
    "        \n",
    "    for k, sm in rec.samples.items() :\n",
    "        # For each samples count which samples have the same alleles tuple --> same variants found at same position \n",
    "        sm[\"SH\"] = sum(al == allAlleles[k] for al in allAlleles.values()) - 1 # minus 1 because do not count itself\n",
    "        \n",
    "    # Create a list with the number of times an allele is shared\n",
    "    # When an allele is found in a sample the cAlleles key is incremented by 1 \n",
    "    sh = []\n",
    "    for al in rec.alleles :\n",
    "        sh.append(cAlleles[al])\n",
    "    rec.info[\"SH\"] = sh # Set the SH tag in the current record for later output to VCF\n",
    "    \n",
    "    # Write out the record\n",
    "    try :\n",
    "        vcf_out.write(rec)\n",
    "    except :\n",
    "        print(\"Here: \", n)\n",
    "        print(rec, \"\\n\")\n",
    "        \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if len(rec.alleles) > 1 : # DEBUGGING\n",
    "        try :\n",
    "            out = \"\\n\".join(\"Sample({}: {} {} {})\".format(k, \"/\".join(str(x) for x in s[\"GT\"]), s[\"SH\"], s[\"TP\"]) for k, s in rec.samples.items())\n",
    "        except :\n",
    "            print(rec)\n",
    "    \"\"\"\n",
    "\n",
    "vcf_in.close()\n",
    "vcf_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Micro-evolution experiment\n",
    "\n",
    "## Create new VCF with SH, TP, GET infos\n",
    "\n",
    "### All samples together (50 and 100 gens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vcf = \"/media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/genotype_allsamples/merged_allsites.vcf\"\n",
    "out_vcf = \"/media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/genotype_allsamples/merged_allsites.GETS.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##ALT=<ID=NON_REF,Description=\"Represents any possible alternative allele not already represented at this location by REF and ALT\">\n",
      "##FILTER=<ID=LowQual,Description=\"Low quality\">\n",
      "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=PGT,Number=1,Type=String,Description=\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles\">\n",
      "##FORMAT=<ID=PID,Number=1,Type=String,Description=\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\">\n",
      "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "##FORMAT=<ID=PS,Number=1,Type=Integer,Description=\"Phasing set (typically the position of the first variant in the set)\">\n",
      "##FORMAT=<ID=RGQ,Number=1,Type=Integer,Description=\"Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong)\">\n",
      "##FORMAT=<ID=SB,Number=4,Type=Integer,Description=\"Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.\">\n",
      "##GATKCommandLine=<ID=GenomicsDBImport,CommandLine=\"GenomicsDBImport --genomicsdb-workspace-path /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/genotype_allsamples/combined.db --batch-size 50 --sample-name-map /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/genotype_allsamples/sample.map.txt --max-num-intervals-to-import-in-parallel 4 --intervals /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/genotype_allsamples/intervals.list --reference /media/urbe/MyBDrive1/Antoine/27-10-21_VariantCalling_MA/input/reference.fa --genomicsdb-segment-size 1048576 --genomicsdb-vcf-buffer-size 16384 --overwrite-existing-genomicsdb-workspace false --consolidate false --validate-sample-name-map false --merge-input-intervals false --reader-threads 1 --merge-contigs-into-num-partitions 0 --genomicsdb-shared-posixfs-optimizations false --genomicsdb-use-gcs-hdfs-connector false --interval-set-rule UNION --interval-padding 0 --interval-exclusion-padding 0 --interval-merging-rule ALL --read-validation-stringency SILENT --seconds-between-progress-updates 10.0 --disable-sequence-dictionary-validation false --create-output-bam-index true --create-output-bam-md5 false --create-output-variant-index true --create-output-variant-md5 false --max-variants-per-shard 0 --lenient false --add-output-sam-program-record true --add-output-vcf-command-line true --cloud-prefetch-buffer 0 --cloud-index-prefetch-buffer 0 --disable-bam-index-caching false --sites-only-vcf-output false --help false --version false --showHidden false --verbosity INFO --QUIET false --use-jdk-deflater false --use-jdk-inflater false --gcs-max-retries 20 --gcs-project-for-requester-pays  --disable-tool-default-read-filters false\",Version=\"4.2.1.0\",Date=\"August 25, 2023 3:12:20 PM CEST\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">\n",
      "##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">\n",
      "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
      "##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">\n",
      "##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">\n",
      "##INFO=<ID=MQRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\">\n",
      "##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant Confidence/Quality by Depth\">\n",
      "##INFO=<ID=RAW_MQandDP,Number=2,Type=Integer,Description=\"Raw data (sum of squared MQ and total depth) for improved RMS Mapping Quality calculation. Incompatible with deprecated RAW_MQ formulation.\">\n",
      "##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\">\n",
      "##INFO=<ID=SOR,Number=1,Type=Float,Description=\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\">\n",
      "##contig=<ID=Chrom_3,length=20354777>\n",
      "##contig=<ID=Chrom_1,length=18146847>\n",
      "##contig=<ID=Chrom_5,length=16930519>\n",
      "##contig=<ID=Chrom_2,length=16274841>\n",
      "##contig=<ID=Chrom_4,length=15224634>\n",
      "##contig=<ID=Chrom_6,length=13893210>\n",
      "##source=GenomicsDBImport\n",
      "##FORMAT=<ID=GET,Number=1,Type=String,Description=\"Type of variation compared to the ancestor sample.\">\n",
      "##FORMAT=<ID=SH,Number=1,Type=Integer,Description=\"Number of samples sharing the same variant as the considered sample.\">\n",
      "##FORMAT=<ID=TP,Number=1,Type=String,Description=\"Type of variant.\">\n",
      "##INFO=<ID=SH,Number=R,Type=Integer,Description=\"Number of samples sharing the same allele.\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tD2A1\tD2B3\tD2B3_50G\tD2C1\tD2C1_50G\tD2C3\tD2C3_50G\tD3A1\tD3A3\tD3A3_50G\tD4A3\tD4B4\tD4B4_50G\tD5B3\tD5C1\tD5C1_50G\tD5C3\tD5C3_50G\tH2A3\tH2B4\tH2B4_50G\tH2C3\tH3A4\tH3A4_50G\tH3C4\tH3C4_50G\tH4A4\tH4A4_50G\tH4C2\tH5A2\tH5A2_50G\tH5A3\tH5A4\tH5A4_50G\tH5C2\tH5C2_50G\tancestor\n",
      "\n",
      "Elapsed records: 0\n",
      "Elapsed records: 1000000\n",
      "Elapsed records: 2000000\n",
      "Elapsed records: 3000000\n",
      "Elapsed records: 4000000\n",
      "Elapsed records: 5000000\n",
      "Elapsed records: 6000000\n",
      "Elapsed records: 7000000\n",
      "Elapsed records: 8000000\n",
      "Elapsed records: 9000000\n",
      "Elapsed records: 10000000\n",
      "Elapsed records: 11000000\n",
      "Elapsed records: 12000000\n",
      "Elapsed records: 13000000\n",
      "Elapsed records: 14000000\n",
      "Elapsed records: 15000000\n",
      "Elapsed records: 16000000\n",
      "Elapsed records: 17000000\n",
      "Elapsed records: 18000000\n",
      "Elapsed records: 19000000\n",
      "Elapsed records: 20000000\n",
      "Elapsed records: 21000000\n",
      "Elapsed records: 22000000\n",
      "Elapsed records: 23000000\n",
      "Elapsed records: 24000000\n",
      "Elapsed records: 25000000\n",
      "Elapsed records: 26000000\n",
      "Elapsed records: 27000000\n",
      "Elapsed records: 28000000\n",
      "Elapsed records: 29000000\n",
      "Elapsed records: 30000000\n",
      "Elapsed records: 31000000\n",
      "Elapsed records: 32000000\n",
      "Elapsed records: 33000000\n",
      "Elapsed records: 34000000\n",
      "Elapsed records: 35000000\n",
      "Elapsed records: 36000000\n",
      "Elapsed records: 37000000\n",
      "Elapsed records: 38000000\n",
      "Elapsed records: 39000000\n",
      "Elapsed records: 40000000\n",
      "Elapsed records: 41000000\n",
      "Elapsed records: 42000000\n",
      "Elapsed records: 43000000\n",
      "Elapsed records: 44000000\n",
      "Elapsed records: 45000000\n",
      "Elapsed records: 46000000\n",
      "Elapsed records: 47000000\n",
      "Elapsed records: 48000000\n",
      "Elapsed records: 49000000\n",
      "Elapsed records: 50000000\n",
      "Elapsed records: 51000000\n",
      "Elapsed records: 52000000\n",
      "Elapsed records: 53000000\n",
      "Elapsed records: 54000000\n",
      "Elapsed records: 55000000\n",
      "Elapsed records: 56000000\n",
      "Elapsed records: 57000000\n",
      "Elapsed records: 58000000\n",
      "Elapsed records: 59000000\n",
      "Elapsed records: 60000000\n",
      "Elapsed records: 61000000\n",
      "Elapsed records: 62000000\n",
      "Elapsed records: 63000000\n",
      "Elapsed records: 64000000\n",
      "Elapsed records: 65000000\n",
      "Elapsed records: 66000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed records: 67000000\n",
      "Elapsed records: 68000000\n",
      "Elapsed records: 69000000\n",
      "Elapsed records: 70000000\n",
      "Elapsed records: 71000000\n",
      "Elapsed records: 72000000\n",
      "Elapsed records: 73000000\n",
      "Elapsed records: 74000000\n",
      "Elapsed records: 75000000\n",
      "Elapsed records: 76000000\n",
      "Elapsed records: 77000000\n",
      "Elapsed records: 78000000\n",
      "Elapsed records: 79000000\n",
      "Elapsed records: 80000000\n",
      "Elapsed records: 81000000\n",
      "Elapsed records: 82000000\n",
      "Elapsed records: 83000000\n",
      "Elapsed records: 84000000\n",
      "Elapsed records: 85000000\n",
      "Elapsed records: 86000000\n",
      "Elapsed records: 87000000\n",
      "Elapsed records: 88000000\n",
      "Elapsed records: 89000000\n",
      "Elapsed records: 90000000\n",
      "Elapsed records: 91000000\n",
      "Elapsed records: 92000000\n",
      "Elapsed records: 93000000\n",
      "Elapsed records: 94000000\n",
      "Elapsed records: 95000000\n",
      "Elapsed records: 96000000\n",
      "Elapsed records: 97000000\n",
      "Elapsed records: 98000000\n",
      "Elapsed records: 99000000\n",
      "Elapsed records: 100000000\n",
      "CPU times: user 18h 46min 14s, sys: 5min 1s, total: 18h 51min 16s\n",
      "Wall time: 18h 52min 25s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "vcf_in = VariantFile(in_vcf)  # auto-detect input format\n",
    "vcf_out = VariantFile(out_vcf, 'w', header=vcf_in.header)\n",
    "\n",
    "# For writing out in a new VCF file\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_out.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele.\")])\n",
    "\n",
    "# Necessary to allow setting new tags in the variant records and record samples\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_in.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele (ref or alt).\")])\n",
    "\n",
    "# DEBUG : check if header was modified indeed\n",
    "print(vcf_out.header)\n",
    "\n",
    "ancestor = \"ancestor\"\n",
    "\n",
    "# For record in sample\n",
    "for n, rec in enumerate(vcf_in) :\n",
    "    if n % 1000000 == 0 :\n",
    "        print(\"Elapsed records: {}\".format(n))\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 11196 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    cAlleles = {k:0 for k in rec.alleles} # For INFO SH\n",
    "    allAlleles = {} # Will store tuples for each sample for later counting\n",
    "    \n",
    "    # For sample in record\n",
    "    for k, sm in rec.samples.items() :\n",
    "        cCurrentAlleles = []\n",
    "            \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 11196 :\n",
    "            print(k, cCurrentAlleles)\n",
    "        \"\"\"\n",
    "           \n",
    "        try : # In case no AD -> go to except\n",
    "            for al, dp in zip(rec.alleles, sm[\"AD\"]) :\n",
    "                # Check each allele if 0.1 < AF < 0.9 or if called by GATK --> add to the cAlleles and cCurrentAlleles\n",
    "                freq = 0.0\n",
    "                try :\n",
    "                    freq = dp/sm[\"DP\"]\n",
    "                except :\n",
    "                    freq = dp/sum(sm[\"AD\"])\n",
    "\n",
    "                if freq >= 0.1 or al in sm.alleles :\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "        except :\n",
    "            try : # In case no DP --> except\n",
    "                if len(set(sm.alleles)) == 1 : # If is homozygous\n",
    "                    al = set(sm.alleles).pop()\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] = 1\n",
    "                    else :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "            except :\n",
    "                print(rec)\n",
    "                print(\"Unexpected error:\", sys.exc_info()[0])\n",
    "                #break\n",
    "                pass\n",
    "\n",
    "        allAlleles[k] = tuple(cCurrentAlleles) # Create a tuple (unchangeable and unsortable --> less mistakes)\n",
    "        # Tuple contains all alleles found in the sample (with 0.1 < AF < 0.9 OR called by GATK)\n",
    "        \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 54641 :\n",
    "            print(k, allAlleles[k])\n",
    "        \"\"\"\n",
    "        \n",
    "        if len(allAlleles[k]) == 0 or None in sm.alleles : # If no alleles found or uncalled genotype\n",
    "            sm[\"TP\"] = \"UC\" # TYPE = Uncalled\n",
    "        else : # At least one allele found\n",
    "            if len(allAlleles[k]) == 1 : # Mono-allelic\n",
    "                if allAlleles[k][0] == rec.ref :\n",
    "                    sm[\"TP\"] = \"RF\" # TYPE = Hom Ref \n",
    "                else :\n",
    "                    if len(allAlleles[k][0]) == 1 and allAlleles[k][0] != \"*\" :\n",
    "                        sm[\"TP\"] = \"AS\" # TYPE = Alternative SNV\n",
    "                    else :\n",
    "                        if allAlleles[k][0] == \"<NON_REF>\" :\n",
    "                            sm[\"TP\"] = \"AU\" # TYPE = Alternative UNKNOWN\n",
    "                        else :\n",
    "                            sm[\"TP\"] = \"AI\" # Alternative INDEL\n",
    "            elif len(allAlleles[k]) == 2 : # TYPE = Bi-allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BS\" # Bi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BU\" # TYPE = Bi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"BI\" # TYPE = Bi-allelic INDEL\n",
    "            else : # Multi-Allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MS\" # TYPE = Multi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MU\" # TYPE = Multi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"MI\" # TYPE = Multi-allelic INDEL\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    try :\n",
    "        Anc = rec.samples[ancestor][\"TP\"]\n",
    "    except :\n",
    "        print(\"NO TP\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "\n",
    "    try :\n",
    "        AncAl = allAlleles[ancestor]\n",
    "    except :\n",
    "        print(\"NO KEYS\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # Add tag GET to each sample\n",
    "    typeDict = {\"UC\":0, \"AU\":0, \"BU\":0, \"MU\":0, \"RF\":1, \"AS\":1, \"AI\": 1, \"BI\":2, \"BS\":2, \"MI\":3, \"MS\":3}\n",
    "    AncAl = allAlleles[ancestor]\n",
    "    Anc = rec.samples[ancestor][\"TP\"]\n",
    "    A = typeDict[Anc]\n",
    "    for k, sm in rec.samples.items() :\n",
    "        if sm.name == ancestor :\n",
    "            continue\n",
    "        else :\n",
    "            OffAl = allAlleles[k]\n",
    "            Off = sm[\"TP\"]\n",
    "            O = typeDict[Off]\n",
    "            \n",
    "            intersect = set(AncAl) & set(OffAl)\n",
    "            NumShared = len(intersect)\n",
    "            AllShared = bool(NumShared == len(set(AncAl))) # If all ancestors alleles are found in the offspring (and vice-versa)\n",
    "            NoneShared = bool(NumShared == 0) # If no common alleles\n",
    "            \n",
    "            if A == 0 or O == 0 :\n",
    "                sm[\"GET\"] = \"U\"\n",
    "            elif A == 1 and O == 1 :\n",
    "                if AllShared : # Same allele --> A or 0\n",
    "                    if Off == \"RF\" : # Ref --> 0\n",
    "                        sm[\"GET\"] = \"0\"\n",
    "                    else : # Not ref --> A\n",
    "                        sm[\"GET\"] = \"A\" \n",
    "                else : # Different allele --> S1\n",
    "                    sm[\"GET\"] = \"S1\"\n",
    "            elif A == 1 and O == 2 : # Offspring heterozygous\n",
    "                if NoneShared : # No common allele --> S4\n",
    "                    sm[\"GET\"] = \"S4\"\n",
    "                else : # 1 shared (cannot be more than one because 2 = biallelic) --> N\n",
    "                    sm[\"GET\"] = \"N\"\n",
    "            elif A == 1 and O == 3 : # Offspring multi-allelic --> always suspect S4\n",
    "                sm[\"GET\"] = \"S4\"\n",
    "            elif A == 2 and O == 1 : # Ancestor heterozygous and offspring homozygous\n",
    "                if NoneShared : # No allele shared --> suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1 allele in common --> cannot be more\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 2 and O == 2 : # both heterozygous\n",
    "                if NoneShared : # No common allele --> S2\n",
    "                    sm[\"GET\"] = \"S2\"\n",
    "                else : \n",
    "                    if NumShared == 1 : # One in common\n",
    "                        sm[\"GET\"] = \"M\"\n",
    "                    else : # Two in common --> identical site <-> AllShared\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "            elif A == 2 and O == 3 : # Ancestor biallelic het and offspring multiallelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S2\" \n",
    "                else :\n",
    "                    if NumShared == 1 : # 1 common allele --> more than 2 events involved : suspect\n",
    "                        sm[\"GET\"] = \"S2\"\n",
    "                    else : # 2 common alleles --> Increased heterozygosity\n",
    "                        sm[\"GET\"] = \"I\"\n",
    "            elif A == 3 and O == 1 : # Ancestor multi-allelic and offspring homozygous\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\" \n",
    "                else : # 1 shared --> lost heterozygosity\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 3 and O == 2 : # Ancestor multi-allelic and offspring bi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    sm[\"GET\"] = \"D\"\n",
    "            elif A == 3 and O == 3 : # Both multi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    if AllShared :\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "                    else :\n",
    "                        if len(AncAl) > len(OffAl) : # More alleles in ancestor : Decreased Het\n",
    "                            sm[\"GET\"] = \"D\"\n",
    "                        else : # More alleles in offspring : Increased Het\n",
    "                            sm[\"GET\"] = \"I\"\n",
    "            else : # Should not come here\n",
    "                # DEBUG\n",
    "                print(n, allAlleles)\n",
    "                print(k, rec)\n",
    "                print(\"----\")\n",
    "                \n",
    "                pass\n",
    "    \n",
    "    # Add UNCALLED samples to the dictionary for later sharedness\n",
    "    for sample in vcf_out.header.samples :\n",
    "        if sample not in allAlleles.keys() :\n",
    "            allAlleles[sample] = tuple([])\n",
    "        \n",
    "    for k, sm in rec.samples.items() :\n",
    "        # For each samples count which samples have the same alleles tuple --> same variants found at same position \n",
    "        sm[\"SH\"] = sum(al == allAlleles[k] for al in allAlleles.values()) - 1 # minus 1 because do not count itself\n",
    "        \n",
    "    # Create a list with the number of times an allele is shared\n",
    "    # When an allele is found in a sample the cAlleles key is incremented by 1 \n",
    "    sh = []\n",
    "    for al in rec.alleles :\n",
    "        sh.append(cAlleles[al])\n",
    "    rec.info[\"SH\"] = sh # Set the SH tag in the current record for later output to VCF\n",
    "    \n",
    "    # Write out the record\n",
    "    try :\n",
    "        vcf_out.write(rec)\n",
    "    except :\n",
    "        print(\"Here: \", n)\n",
    "        print(rec, \"\\n\")\n",
    "        \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if len(rec.alleles) > 1 : # DEBUGGING\n",
    "        try :\n",
    "            out = \"\\n\".join(\"Sample({}: {} {} {})\".format(k, \"/\".join(str(x) for x in s[\"GT\"]), s[\"SH\"], s[\"TP\"]) for k, s in rec.samples.items())\n",
    "        except :\n",
    "            print(rec)\n",
    "    \"\"\"\n",
    "\n",
    "vcf_in.close()\n",
    "vcf_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome Repair experiment\n",
    "\n",
    "## Create new VCF with SH, TP, GET infos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vcf = \"/media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/jointgenotyping/merged_allsites.vcf\"\n",
    "out_vcf = \"/media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/jointgenotyping/merged_allsites.GETS.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##ALT=<ID=NON_REF,Description=\"Represents any possible alternative allele not already represented at this location by REF and ALT\">\n",
      "##FILTER=<ID=LowQual,Description=\"Low quality\">\n",
      "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=PGT,Number=1,Type=String,Description=\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles\">\n",
      "##FORMAT=<ID=PID,Number=1,Type=String,Description=\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\">\n",
      "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "##FORMAT=<ID=PS,Number=1,Type=Integer,Description=\"Phasing set (typically the position of the first variant in the set)\">\n",
      "##FORMAT=<ID=RGQ,Number=1,Type=Integer,Description=\"Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong)\">\n",
      "##FORMAT=<ID=SB,Number=4,Type=Integer,Description=\"Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.\">\n",
      "##GATKCommandLine=<ID=GenomicsDBImport,CommandLine=\"GenomicsDBImport --genomicsdb-workspace-path /media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/jointgenotyping/combined.db --batch-size 50 --sample-name-map /media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/jointgenotyping/sample.map.txt --max-num-intervals-to-import-in-parallel 4 --intervals /media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/jointgenotyping/intervals.list --reference /media/urbe/MyADrive1/Antoine/19-11-21_VariantCalling_ARC/input/reference.fa --genomicsdb-segment-size 1048576 --genomicsdb-vcf-buffer-size 16384 --overwrite-existing-genomicsdb-workspace false --consolidate false --validate-sample-name-map false --merge-input-intervals false --reader-threads 1 --merge-contigs-into-num-partitions 0 --genomicsdb-shared-posixfs-optimizations false --genomicsdb-use-gcs-hdfs-connector false --interval-set-rule UNION --interval-padding 0 --interval-exclusion-padding 0 --interval-merging-rule ALL --read-validation-stringency SILENT --seconds-between-progress-updates 10.0 --disable-sequence-dictionary-validation false --create-output-bam-index true --create-output-bam-md5 false --create-output-variant-index true --create-output-variant-md5 false --max-variants-per-shard 0 --lenient false --add-output-sam-program-record true --add-output-vcf-command-line true --cloud-prefetch-buffer 0 --cloud-index-prefetch-buffer 0 --disable-bam-index-caching false --sites-only-vcf-output false --help false --version false --showHidden false --verbosity INFO --QUIET false --use-jdk-deflater false --use-jdk-inflater false --gcs-max-retries 20 --gcs-project-for-requester-pays  --disable-tool-default-read-filters false\",Version=\"4.2.1.0\",Date=\"August 28, 2023 8:08:24 AM CEST\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">\n",
      "##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">\n",
      "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
      "##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">\n",
      "##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">\n",
      "##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">\n",
      "##INFO=<ID=MQRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\">\n",
      "##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant Confidence/Quality by Depth\">\n",
      "##INFO=<ID=RAW_MQandDP,Number=2,Type=Integer,Description=\"Raw data (sum of squared MQ and total depth) for improved RMS Mapping Quality calculation. Incompatible with deprecated RAW_MQ formulation.\">\n",
      "##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\">\n",
      "##INFO=<ID=SOR,Number=1,Type=Float,Description=\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\">\n",
      "##contig=<ID=Chrom_3,length=20354777>\n",
      "##contig=<ID=Chrom_1,length=18146847>\n",
      "##contig=<ID=Chrom_5,length=16930519>\n",
      "##contig=<ID=Chrom_2,length=16274841>\n",
      "##contig=<ID=Chrom_4,length=15224634>\n",
      "##contig=<ID=Chrom_6,length=13893210>\n",
      "##source=GenomicsDBImport\n",
      "##FORMAT=<ID=GET,Number=1,Type=String,Description=\"Type of variation compared to the ancestor sample.\">\n",
      "##FORMAT=<ID=SH,Number=1,Type=Integer,Description=\"Number of samples sharing the same variant as the considered sample.\">\n",
      "##FORMAT=<ID=TP,Number=1,Type=String,Description=\"Type of variant.\">\n",
      "##INFO=<ID=SH,Number=R,Type=Integer,Description=\"Number of samples sharing the same allele.\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\t30D_C13_E3\t30D_C38_E4\t30D_C38_E5\t30D_C52_E5\t30H_C36_E5\t30H_C3_E4\t30H_C3_E5\t30H_C48_E5\tP0_C27_E5\tP0_C40_E5\tP0_C9_E4\tP0_C9_E5\tP100_C30_E3\tP100_C30_E4\tP100_C8_E3\tP100_C8_E4\tP250_C17_E3\tP250_C17_E4\tP250_C8_E3\tP250_C8_E4\tP500_C16_E4\tP500_C16_E5\tP500_C18_E3\tP500_C30_E3\tancestor\n",
      "\n",
      "Elapsed records: 0\n",
      "Elapsed records: 1000000\n",
      "Elapsed records: 2000000\n",
      "Elapsed records: 3000000\n",
      "Elapsed records: 4000000\n",
      "Elapsed records: 5000000\n",
      "Elapsed records: 6000000\n",
      "Elapsed records: 7000000\n",
      "Elapsed records: 8000000\n",
      "Elapsed records: 9000000\n",
      "Elapsed records: 10000000\n",
      "Elapsed records: 11000000\n",
      "Elapsed records: 12000000\n",
      "Elapsed records: 13000000\n",
      "Elapsed records: 14000000\n",
      "Elapsed records: 15000000\n",
      "Elapsed records: 16000000\n",
      "Elapsed records: 17000000\n",
      "Elapsed records: 18000000\n",
      "Elapsed records: 19000000\n",
      "Elapsed records: 20000000\n",
      "Elapsed records: 21000000\n",
      "Elapsed records: 22000000\n",
      "Elapsed records: 23000000\n",
      "Elapsed records: 24000000\n",
      "Elapsed records: 25000000\n",
      "Elapsed records: 26000000\n",
      "Elapsed records: 27000000\n",
      "Elapsed records: 28000000\n",
      "Elapsed records: 29000000\n",
      "Elapsed records: 30000000\n",
      "Elapsed records: 31000000\n",
      "Elapsed records: 32000000\n",
      "Elapsed records: 33000000\n",
      "Elapsed records: 34000000\n",
      "Elapsed records: 35000000\n",
      "Elapsed records: 36000000\n",
      "Elapsed records: 37000000\n",
      "Elapsed records: 38000000\n",
      "Elapsed records: 39000000\n",
      "Elapsed records: 40000000\n",
      "Elapsed records: 41000000\n",
      "Elapsed records: 42000000\n",
      "Elapsed records: 43000000\n",
      "Elapsed records: 44000000\n",
      "Elapsed records: 45000000\n",
      "Elapsed records: 46000000\n",
      "Elapsed records: 47000000\n",
      "Elapsed records: 48000000\n",
      "Elapsed records: 49000000\n",
      "Elapsed records: 50000000\n",
      "Elapsed records: 51000000\n",
      "Elapsed records: 52000000\n",
      "Elapsed records: 53000000\n",
      "Elapsed records: 54000000\n",
      "Elapsed records: 55000000\n",
      "Elapsed records: 56000000\n",
      "Elapsed records: 57000000\n",
      "Elapsed records: 58000000\n",
      "Elapsed records: 59000000\n",
      "Elapsed records: 60000000\n",
      "Elapsed records: 61000000\n",
      "Elapsed records: 62000000\n",
      "Elapsed records: 63000000\n",
      "Elapsed records: 64000000\n",
      "Elapsed records: 65000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed records: 66000000\n",
      "Elapsed records: 67000000\n",
      "Elapsed records: 68000000\n",
      "Elapsed records: 69000000\n",
      "Elapsed records: 70000000\n",
      "Elapsed records: 71000000\n",
      "Elapsed records: 72000000\n",
      "Elapsed records: 73000000\n",
      "Elapsed records: 74000000\n",
      "Elapsed records: 75000000\n",
      "Elapsed records: 76000000\n",
      "Elapsed records: 77000000\n",
      "Elapsed records: 78000000\n",
      "Elapsed records: 79000000\n",
      "Elapsed records: 80000000\n",
      "Elapsed records: 81000000\n",
      "Elapsed records: 82000000\n",
      "Elapsed records: 83000000\n",
      "Elapsed records: 84000000\n",
      "Elapsed records: 85000000\n",
      "Elapsed records: 86000000\n",
      "Elapsed records: 87000000\n",
      "Elapsed records: 88000000\n",
      "Elapsed records: 89000000\n",
      "Elapsed records: 90000000\n",
      "Elapsed records: 91000000\n",
      "Elapsed records: 92000000\n",
      "Elapsed records: 93000000\n",
      "Elapsed records: 94000000\n",
      "Elapsed records: 95000000\n",
      "Elapsed records: 96000000\n",
      "Elapsed records: 97000000\n",
      "Elapsed records: 98000000\n",
      "Elapsed records: 99000000\n",
      "Elapsed records: 100000000\n",
      "CPU times: user 11h 25min 39s, sys: 3min 15s, total: 11h 28min 54s\n",
      "Wall time: 11h 29min 25s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "vcf_in = VariantFile(in_vcf)  # auto-detect input format\n",
    "vcf_out = VariantFile(out_vcf, 'w', header=vcf_in.header)\n",
    "\n",
    "# For writing out in a new VCF file\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_out.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_out.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele.\")])\n",
    "\n",
    "# Necessary to allow setting new tags in the variant records and record samples\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"GET\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variation compared to the ancestor sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"SH\"), (\"Number\", \"1\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same variant as the considered sample.\")])\n",
    "vcf_in.header.add_meta(key=\"FORMAT\", items=[(\"ID\", \"TP\"), (\"Number\", \"1\"), (\"Type\", \"String\"), (\"Description\", \"Type of variant.\")])\n",
    "vcf_in.header.add_meta(key=\"INFO\", items=[(\"ID\", \"SH\"), (\"Number\", \"R\"), (\"Type\", \"Integer\"), (\"Description\", \"Number of samples sharing the same allele (ref or alt).\")])\n",
    "\n",
    "# DEBUG : check if header was modified indeed\n",
    "print(vcf_out.header)\n",
    "\n",
    "ancestor = \"ancestor\"\n",
    "\n",
    "# For record in sample\n",
    "for n, rec in enumerate(vcf_in) :\n",
    "    if n % 1000000 == 0 :\n",
    "        print(\"Elapsed records: {}\".format(n))\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 11196 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    cAlleles = {k:0 for k in rec.alleles} # For INFO SH\n",
    "    allAlleles = {} # Will store tuples for each sample for later counting\n",
    "    \n",
    "    # For sample in record\n",
    "    for k, sm in rec.samples.items() :\n",
    "        cCurrentAlleles = []\n",
    "            \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 11196 :\n",
    "            print(k, cCurrentAlleles)\n",
    "        \"\"\"\n",
    "           \n",
    "        try : # In case no AD -> go to except\n",
    "            for al, dp in zip(rec.alleles, sm[\"AD\"]) :\n",
    "                # Check each allele if 0.1 < AF < 0.9 or if called by GATK --> add to the cAlleles and cCurrentAlleles\n",
    "                freq = 0.0\n",
    "                try :\n",
    "                    freq = dp/sm[\"DP\"]\n",
    "                except :\n",
    "                    freq = dp/sum(sm[\"AD\"])\n",
    "\n",
    "                if freq >= 0.1 or al in sm.alleles :\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "        except :\n",
    "            try : # In case no DP --> except\n",
    "                if len(set(sm.alleles)) == 1 : # If is homozygous\n",
    "                    al = set(sm.alleles).pop()\n",
    "                    if al not in cAlleles :\n",
    "                        cAlleles[al] = 1\n",
    "                    else :\n",
    "                        cAlleles[al] += 1\n",
    "                    if al not in cCurrentAlleles :\n",
    "                        cCurrentAlleles.append(al)\n",
    "            except :\n",
    "                print(rec)\n",
    "                print(\"Unexpected error:\", sys.exc_info()[0])\n",
    "                #break\n",
    "                pass\n",
    "\n",
    "        allAlleles[k] = tuple(cCurrentAlleles) # Create a tuple (unchangeable and unsortable --> less mistakes)\n",
    "        # Tuple contains all alleles found in the sample (with 0.1 < AF < 0.9 OR called by GATK)\n",
    "        \n",
    "        # DEBUG\n",
    "        \"\"\"\n",
    "        if rec.pos == 54641 :\n",
    "            print(k, allAlleles[k])\n",
    "        \"\"\"\n",
    "        \n",
    "        if len(allAlleles[k]) == 0 or None in sm.alleles : # If no alleles found or uncalled genotype\n",
    "            sm[\"TP\"] = \"UC\" # TYPE = Uncalled\n",
    "        else : # At least one allele found\n",
    "            if len(allAlleles[k]) == 1 : # Mono-allelic\n",
    "                if allAlleles[k][0] == rec.ref :\n",
    "                    sm[\"TP\"] = \"RF\" # TYPE = Hom Ref \n",
    "                else :\n",
    "                    if len(allAlleles[k][0]) == 1 and allAlleles[k][0] != \"*\" :\n",
    "                        sm[\"TP\"] = \"AS\" # TYPE = Alternative SNV\n",
    "                    else :\n",
    "                        if allAlleles[k][0] == \"<NON_REF>\" :\n",
    "                            sm[\"TP\"] = \"AU\" # TYPE = Alternative UNKNOWN\n",
    "                        else :\n",
    "                            sm[\"TP\"] = \"AI\" # Alternative INDEL\n",
    "            elif len(allAlleles[k]) == 2 : # TYPE = Bi-allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BS\" # Bi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"BU\" # TYPE = Bi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"BI\" # TYPE = Bi-allelic INDEL\n",
    "            else : # Multi-Allelic\n",
    "                if all(len(x) == 1 for x in allAlleles[k]) and \"*\" not in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MS\" # TYPE = Multi-allelic SNV\n",
    "                elif \"<NON_REF>\" in allAlleles[k] :\n",
    "                    sm[\"TP\"] = \"MU\" # TYPE = Multi-allelic UNKNOWN\n",
    "                else :\n",
    "                    sm[\"TP\"] = \"MI\" # TYPE = Multi-allelic INDEL\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    try :\n",
    "        Anc = rec.samples[ancestor][\"TP\"]\n",
    "    except :\n",
    "        print(\"NO TP\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "\n",
    "    try :\n",
    "        AncAl = allAlleles[ancestor]\n",
    "    except :\n",
    "        print(\"NO KEYS\")\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(allAlleles)\n",
    "        print(n, rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # Add tag GET to each sample\n",
    "    typeDict = {\"UC\":0, \"AU\":0, \"BU\":0, \"MU\":0, \"RF\":1, \"AS\":1, \"AI\": 1, \"BI\":2, \"BS\":2, \"MI\":3, \"MS\":3}\n",
    "    AncAl = allAlleles[ancestor]\n",
    "    Anc = rec.samples[ancestor][\"TP\"]\n",
    "    A = typeDict[Anc]\n",
    "    for k, sm in rec.samples.items() :\n",
    "        if sm.name == ancestor :\n",
    "            continue\n",
    "        else :\n",
    "            OffAl = allAlleles[k]\n",
    "            Off = sm[\"TP\"]\n",
    "            O = typeDict[Off]\n",
    "            \n",
    "            intersect = set(AncAl) & set(OffAl)\n",
    "            NumShared = len(intersect)\n",
    "            AllShared = bool(NumShared == len(set(AncAl))) # If all ancestors alleles are found in the offspring (and vice-versa)\n",
    "            NoneShared = bool(NumShared == 0) # If no common alleles\n",
    "            \n",
    "            if A == 0 or O == 0 :\n",
    "                sm[\"GET\"] = \"U\"\n",
    "            elif A == 1 and O == 1 :\n",
    "                if AllShared : # Same allele --> A or 0\n",
    "                    if Off == \"RF\" : # Ref --> 0\n",
    "                        sm[\"GET\"] = \"0\"\n",
    "                    else : # Not ref --> A\n",
    "                        sm[\"GET\"] = \"A\" \n",
    "                else : # Different allele --> S1\n",
    "                    sm[\"GET\"] = \"S1\"\n",
    "            elif A == 1 and O == 2 : # Offspring heterozygous\n",
    "                if NoneShared : # No common allele --> S4\n",
    "                    sm[\"GET\"] = \"S4\"\n",
    "                else : # 1 shared (cannot be more than one because 2 = biallelic) --> N\n",
    "                    sm[\"GET\"] = \"N\"\n",
    "            elif A == 1 and O == 3 : # Offspring multi-allelic --> always suspect S4\n",
    "                sm[\"GET\"] = \"S4\"\n",
    "            elif A == 2 and O == 1 : # Ancestor heterozygous and offspring homozygous\n",
    "                if NoneShared : # No allele shared --> suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1 allele in common --> cannot be more\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 2 and O == 2 : # both heterozygous\n",
    "                if NoneShared : # No common allele --> S2\n",
    "                    sm[\"GET\"] = \"S2\"\n",
    "                else : \n",
    "                    if NumShared == 1 : # One in common\n",
    "                        sm[\"GET\"] = \"M\"\n",
    "                    else : # Two in common --> identical site <-> AllShared\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "            elif A == 2 and O == 3 : # Ancestor biallelic het and offspring multiallelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S2\" \n",
    "                else :\n",
    "                    if NumShared == 1 : # 1 common allele --> more than 2 events involved : suspect\n",
    "                        sm[\"GET\"] = \"S2\"\n",
    "                    else : # 2 common alleles --> Increased heterozygosity\n",
    "                        sm[\"GET\"] = \"I\"\n",
    "            elif A == 3 and O == 1 : # Ancestor multi-allelic and offspring homozygous\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\" \n",
    "                else : # 1 shared --> lost heterozygosity\n",
    "                    sm[\"GET\"] = \"L\"\n",
    "            elif A == 3 and O == 2 : # Ancestor multi-allelic and offspring bi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    sm[\"GET\"] = \"D\"\n",
    "            elif A == 3 and O == 3 : # Both multi-allelic\n",
    "                if NoneShared : # No common allele --> More than 2 events involved : suspect\n",
    "                    sm[\"GET\"] = \"S3\"\n",
    "                else : # 1+ shared\n",
    "                    if AllShared :\n",
    "                        sm[\"GET\"] = \"C\"\n",
    "                    else :\n",
    "                        if len(AncAl) > len(OffAl) : # More alleles in ancestor : Decreased Het\n",
    "                            sm[\"GET\"] = \"D\"\n",
    "                        else : # More alleles in offspring : Increased Het\n",
    "                            sm[\"GET\"] = \"I\"\n",
    "            else : # Should not come here\n",
    "                # DEBUG\n",
    "                print(n, allAlleles)\n",
    "                print(k, rec)\n",
    "                print(\"----\")\n",
    "                \n",
    "                pass\n",
    "    \n",
    "    # Add UNCALLED samples to the dictionary for later sharedness\n",
    "    for sample in vcf_out.header.samples :\n",
    "        if sample not in allAlleles.keys() :\n",
    "            allAlleles[sample] = tuple([])\n",
    "        \n",
    "    for k, sm in rec.samples.items() :\n",
    "        # For each samples count which samples have the same alleles tuple --> same variants found at same position \n",
    "        sm[\"SH\"] = sum(al == allAlleles[k] for al in allAlleles.values()) - 1 # minus 1 because do not count itself\n",
    "        \n",
    "    # Create a list with the number of times an allele is shared\n",
    "    # When an allele is found in a sample the cAlleles key is incremented by 1 \n",
    "    sh = []\n",
    "    for al in rec.alleles :\n",
    "        sh.append(cAlleles[al])\n",
    "    rec.info[\"SH\"] = sh # Set the SH tag in the current record for later output to VCF\n",
    "    \n",
    "    # Write out the record\n",
    "    try :\n",
    "        vcf_out.write(rec)\n",
    "    except :\n",
    "        print(\"Here: \", n)\n",
    "        print(rec, \"\\n\")\n",
    "        \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if rec.pos == 1603103 :\n",
    "        print(rec)\n",
    "    \"\"\"\n",
    "    \n",
    "    # DEBUG\n",
    "    \"\"\"\n",
    "    if len(rec.alleles) > 1 : # DEBUGGING\n",
    "        try :\n",
    "            out = \"\\n\".join(\"Sample({}: {} {} {})\".format(k, \"/\".join(str(x) for x in s[\"GT\"]), s[\"SH\"], s[\"TP\"]) for k, s in rec.samples.items())\n",
    "        except :\n",
    "            print(rec)\n",
    "    \"\"\"\n",
    "\n",
    "vcf_in.close()\n",
    "vcf_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
